### PR TITLE
KDF doc: expand on 'salt' and 'info' parameters

### PR DESCRIPTION
**Pull request by DDvO**
* `EVP_KDF.pod:` extend text on 'salt' and 'info' parameters
* `openssl-kdf.pod.in:` add text on 'salt' and 'info' parameters; small further improvements

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

File: doc/man1/openssl-kdf.pod.in [hunk 1]
```diff
@@ -81,9 +81,8 @@ A key must be specified for most KDF algorithms.
 
 =item B<hexkey:>I<string>
 
-Specifies the secret key in hexadecimal form (two hex digits per byte).
-The key length must conform to any restrictions of the KDF algorithm.
-A key must be specified for most KDF algorithms.
+As before, but
```
Comment (slontis) [inline doc/man1/openssl-kdf.pod.in:84]: 'As before' doesnt seem clearer to me..

File: doc/man1/openssl-kdf.pod.in [hunk 2]
```diff
@@ -66,7 +66,7 @@ cases.
 =item B<-kdfopt> I<nm>:I<v>
 
 Passes options to the KDF algorithm.
-A comprehensive list of parameters can be found in the EVP_KDF_CTX
```
Comment (slontis) [inline doc/man1/openssl-kdf.pod.in:69]: Maybe reword to
Refer to L<EVP_KDF(3)/PARAMETERS> for a list of parameters.

File: doc/man1/openssl-kdf.pod.in (hunk 1 above)
Comment (slontis) [reply doc/man1/openssl-kdf.pod.in:84]: Similar to the I<key> option

File: doc/man1/openssl-kdf.pod.in [hunk 3]
```diff
@@ -93,8 +92,32 @@ The password must be specified for PBKDF2 and scrypt.
 
 =item B<hexpass:>I<string>
 
-Specifies the password in hexadecimal form (two hex digits per byte).
-The password must be specified for PBKDF2 and scrypt.
+As before, but
```
Comment (slontis) [inline doc/man1/openssl-kdf.pod.in:95]: Same herr..

File: doc/man1/openssl-kdf.pod.in [hunk 4]
```diff
@@ -93,8 +92,32 @@ The password must be specified for PBKDF2 and scrypt.
...
+specifies the password in hexadecimal form (two hex digits per byte).
+
+=item B<salt:>I<string>
+
+Specifies a non-secret random cryptographic salt as an alphanumeric string
+(use if it contains printable characters only).
+The length must conform to any restrictions of the KDF algorithm.
+A salt parameter is required for several KDF algorithms, such as PBKDF2.
```
Comment (slontis) [inline doc/man1/openssl-kdf.pod.in:103]: Maybe provide the link here L<EVP_KDF-PBKDF2(7)>

File: doc/man1/openssl-kdf.pod.in (hunk 1 above)
Comment (DDvO) [reply doc/man1/openssl-kdf.pod.in]: Good point, changed to `Similar to the B<key:> option`,
and likewise for the other three occurences.

File: doc/man1/openssl-kdf.pod.in (hunk 2 above)
Comment (DDvO) [reply doc/man1/openssl-kdf.pod.in:69]: done

File: doc/man1/openssl-kdf.pod.in (hunk 4 above)
Comment (DDvO) [reply doc/man1/openssl-kdf.pod.in]: done

**Comment by DDvO** [comment]
Thanks for your swift reviews and @slontis for your comments.
I've handled them; @paulidale wanna reapprove?

File: doc/man1/openssl-kdf.pod.in (hunk 2 above)
Comment (slontis) [inline doc/man1/openssl-kdf.pod.in:69]: can be found in L<,,,>.
The implementaion documentation bit is not needed.

File: doc/man3/EVP_KDF.pod [hunk 5]
```diff
@@ -229,7 +229,9 @@ sets the key.
 
 =item "info" (B<OSSL_KDF_PARAM_INFO>) <octet string>
 
-This parameter sets the info, fixed info, other info or shared info value.
+Some KDF implementations such as HKDF take an 'info' parameter for binding
```
Comment (slontis) [inline doc/man3/EVP_KDF.pod:232]: I would also link the 2 refs to HKDF.

File: doc/man1/openssl-kdf.pod.in (hunk 2 above)
Comment (DDvO) [reply doc/man1/openssl-kdf.pod.in:69]: done

File: doc/man3/EVP_KDF.pod (hunk 5 above)
Comment (DDvO) [reply doc/man3/EVP_KDF.pod]: done

File: doc/man1/openssl-kdf.pod.in [hunk 6]
```diff
@@ -81,9 +80,8 @@ A key must be specified for most KDF algorithms.
 
 =item B<hexkey:>I<string>
 
-Specifies the secret key in hexadecimal form (two hex digits per byte).
-The key length must conform to any restrictions of the KDF algorithm.
-A key must be specified for most KDF algorithms.
+Similar to the B<key:> option, but
+specifies the secret key in hexadecimal form (two hex digits per byte).
```
Comment (tom-cosgrove-arm) [inline doc/man1/openssl-kdf.pod.in:84]:
```suggestion
Alternative to the B<key:> option where
the secret key is specified in hexadecimal form (two hex digits per byte).
```

File: doc/man1/openssl-kdf.pod.in [hunk 7]
```diff
@@ -93,8 +91,35 @@ The password must be specified for PBKDF2 and scrypt.
...
+such as L<EVP_KDF-PBKDF2(7)>.
+
+=item B<hexsalt:>I<string>
+
+Similar to the B<salt:> option, but
+specifies the cryptographic salt in hexadecimal form (two hex digits per byte).
+
+=item B<info:>I<string>
+
+Some KDF implementations such as L<EVP_KDF-HKDF(7)> take an 'info' parameter
```
[the remark this answers is not in the document]
Comment (DDvO) [reply doc/man1/openssl-kdf.pod.in]: good thought - done, also in `EVP_KDF.pod`

**Comment by DDvO** [comment]
General question on such doc-only PRs:
Should anyone place the **tests:exempted** label?
For such cases I miss a **tests:not applicable** label.

File: doc/man1/openssl-kdf.pod.in [hunk 8]
```diff
@@ -93,8 +91,35 @@ The password must be specified for PBKDF2 and scrypt.
...
 =item B<hexpass:>I<string>
 
-Specifies the password in hexadecimal form (two hex digits per byte).
-The password must be specified for PBKDF2 and scrypt.
+Alternative to the B<pass:> option where
+the password is specified in hexadecimal form (two hex digits per byte).
+
+=item B<salt:>I<string>
+
+Specifies a non-secret random cryptographic salt as an alphanumeric string
```
Comment (tom-cosgrove-arm) [inline doc/man1/openssl-kdf.pod.in:99]: I worry about the use of `random` and `cryptographic` here.

First, with `random`, this might suggest it should be different each time it is used, as opposed to for each different key we want.

And secondly, while salts should be unique, unpredictability is "merely" "better" - being somewhat predictable (e.g. using a non-cryptographic random number generator) isn't as much of a deal-breaker as it is in the case of a symmetric key.

I wonder if we should use `unique` and `unpredictable` here, rather than `cryptographically random`?

Comment (DDvO) [reply doc/man1/openssl-kdf.pod.in:99]: I'm not sure whether `random` or `unpredictable` is better here - what do others think?

Yet note that the proposed text does not say `cryptographically random`.
I deliberately chose `cryptographic salt` because this term is used on https://en.wikipedia.org/wiki/Key_derivation_function and I like it for distinction from regular salt (spice).
In the way I prepended 'cryptographic' to 'salt', it should be harmless and in IMHO somewhat helpful.

**Comment by t8m** [comment]
> General question on such doc-only PRs: Should anyone place the **tests:exempted** label? For such cases I miss a **tests:not applicable** label.

`tests: exempted` has the exact meaning. Yeah, we do not distinguish between the cases where the test would be unproportionally hard to write and when there is nothing to test here.

File: doc/man1/openssl-kdf.pod.in (hunk 8 above)
Comment (t8m) [reply doc/man1/openssl-kdf.pod.in:99]: `a non-secret unique cryptographic salt` would be my choice

Comment (DDvO) [reply doc/man1/openssl-kdf.pod.in]: Done this now (in both documents).

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks.
